### PR TITLE
fix(vrl): fix `Kind::reduced_kind` when `known` is empty

### PR DESCRIPTION
**Pull request by fuchsnj**
`Kind::reduced_kind` produced the wrong value when the `known` list was empty. This led to some obscure type-def bugs. See the test included for a full example.

### CODE REVIEW

File: lib/vrl/tests/tests/internal/closure_arg_kind.vrl [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# result: [
+#           {
+#             "log_events": {
+#               "id": "35683658089614582423604394983260738922885519999578275840",
+#               "message": "{\"bytes\":26780,\"datetime\":\"14/Sep/2020:11:45:41 -0400\",\"host\":\"157.130.216.193\",\"method\":\"PUT\",\"protocol\":\"HTTP/1.0\",\"referer\":\"https://www.principalcross-platform.io/markets/ubiquitous\",\"request\":\"/expedite/convergence\",\"source_type\":\"stdin\",\"status\":301,\"user-identifier\":\"-\"}",
+#               "timestamp": "2020-09-14T19:09:29.039Z"
+#             },
```
Comment (spencergilbert) [inline lib/vrl/tests/tests/internal/closure_arg_kind.vrl:7]:
```suggestion
#             "id": "35683658089614582423604394983260738922885519999578275840",
#             "message": "{\"bytes\":26780,\"datetime\":\"14/Sep/2020:11:45:41 -0400\",\"host\":\"157.130.216.193\",\"method\":\"PUT\",\"protocol\":\"HTTP/1.0\",\"referer\":\"https://www.principalcross-platform.io/markets/ubiquitous\",\"request\":\"/expedite/convergence\",\"source_type\":\"stdin\",\"status\":301,\"user-identifier\":\"-\"}",
#             "timestamp": "2020-09-14T19:09:29.039Z",
```

Post `map_values` the `.log_events` object should be removed and the contents at the root.

File: lib/vrl/tests/tests/internal/closure_arg_kind.vrl [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+#             "subscription_filters": [
+#               "Destination"
+#             ]
+#           },
+#           {
+#             "log_events": {
+#               "id": "35683658089659183914001456229543810359430816722590236673",
+#               "message": "{\"bytes\":17707,\"datetime\":\"14/Sep/2020:11:45:41 -0400\",\"host\":\"109.81.244.252\",\"method\":\"GET\",\"protocol\":\"HTTP/2.0\",\"referer\":\"http://www.investormission-critical.io/24/7/vortals\",\"request\":\"/scale/functionalities/optimize\",\"source_type\":\"stdin\",\"status\":502,\"user-identifier\":\"feeney1708\"}",
+#               "timestamp": "2020-09-14T19:09:29.041Z"
+#             },
```
Comment (spencergilbert) [inline lib/vrl/tests/tests/internal/closure_arg_kind.vrl:21]:
```suggestion
#             "id": "35683658089659183914001456229543810359430816722590236673",
#             "message": "{\"bytes\":17707,\"datetime\":\"14/Sep/2020:11:45:41 -0400\",\"host\":\"109.81.244.252\",\"method\":\"GET\",\"protocol\":\"HTTP/2.0\",\"referer\":\"http://www.investormission-critical.io/24/7/vortals\",\"request\":\"/scale/functionalities/optimize\",\"source_type\":\"stdin\",\"status\":502,\"user-identifier\":\"feeney1708\"}",
#             "timestamp": "2020-09-14T19:09:29.041Z",
```

File: lib/vrl/tests/tests/internal/closure_arg_kind.vrl [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+parsed = parse_aws_cloudwatch_log_subscription_message!(.message)
+. = unnest(parsed.log_events)
+map_values(.) -> |value| { events = del(value.log_events) ; value |= events }
```
Comment (spencergilbert) [inline lib/vrl/tests/tests/internal/closure_arg_kind.vrl:37]:
```suggestion
map_values(.) -> |value| { event = del(value.log_events) ; value |= event }
```
I also guess, for readability, this should now be a single "event" so dropping the plural..

File: lib/vrl/tests/tests/internal/closure_arg_kind.vrl (hunk 1 above)
Comment (fuchsnj) [reply lib/vrl/tests/tests/internal/closure_arg_kind.vrl:7]: oh, yea. I needed to assign the result of `map_values` back to `.`

**Review by spencergilbert — APPROVED** [review]

**Pull request merged**